### PR TITLE
Redis Heartbeat

### LINKED ISSUE #2773: Redis Heartbeart
Add heartbeat support to the Redis storage backend similar to the RDB storage.

## Motivation

While the RDB storage backend supports the heartbeat functionality, the Redis storage does not. This prevents the Redis backend from being a drop-in replacement for the RDB backend.

## Description

Implement the methods of `BaseStorage` in `RedisStorage`.

## Alternatives

Move the heartbeat functionality out of the individual storage classes into `BaseStorage`, e.g. via trial system attributes. Although this provides some simplifications, I would not recommend this because deleting stale trials can be implemented more efficiently by the backend.

## Additional context

I would like to contribute this feature, if you think it fits into the library.

### PR DESCRIPTION
**Pull request by Turakar**
## Motivation
Fix #2773. Sorry for not waiting for your replies there, but I could not wait to start coding :smile: 

## Description of the changes
This PR adds heartbeat support to the Redis storage backend and unifies some of the heartbeat related tests for RDB and Redis backends. Detailed list of changes:

- Heartbeats are saved in a hash (i.e. dict) per study and not per trial. While this requires an additional lookup of the study ID in `set_trial_state()` and `record_heartbeat()`, it significantly improves the performance of `fail_stale_trials()` for studies with many non-running trials.
- Heartbeats are removed when a trial's state is updated on completion. As empty hashes are deleted in Redis automatically, there should be no permanent storage usage after finishing all running trials. However, a user can call `record_heartbeat` out of order, but as he/she is not supposed to call backend methods anyway and the RDB storage has the same problem, I ignored this. After all, looking up the trial state in `record_heartbeat()` would cause a performance cost.
- I moved the tests from the RDB storage tests to the general storage tests and added a new list `STORAGE_MODES_HEARTBEAT`. Additionally, `StorageSupplier` now supports passing `kwargs` to the backends to unify the setting of `heartbeat_interval`, `grace_period` and `failed_trial_callback`. In this process, I lost one single test which is redundant in my opinion, i.e. checking the return value of `_get_stale_trials()` in `test_fail_stale_trials()`.

### CODE REVIEW

**Comment by Turakar** [comment]
I am not sure about the CI failures, but I think it is not my fault, right?

**Comment by Crissman** [comment]
@keisuke-umezawa Can you take a look at this?

**Review by not522 — COMMENTED** [review]
Sorry for the late response. I left some initial comments. I'll take another look later.

**Review by not522 — COMMENTED** [review]
Basically LGTM. Could you check my comments?

**Comment by not522** [comment]
Could you merge master to fix the CI failure?

File: optuna/storages/_redis.py
Comment (Turakar) [reply optuna/storages/_redis.py:404]: Yes, but it I wanted to make it explicit as I have to do the decoding manually. So while the Redis library automatically converts it to a string, it does not automatically convert it back. But this a matter of personal preference, so I will adjust to your wishes if you want.

Comment (Turakar) [reply optuna/storages/_redis.py:403]: I just checked: Not doing the explicit conversion raises mypy errors.

Comment (not522) [reply optuna/storages/_redis.py:403]: That makes sense. Let's keep explicit conversion.

**Review by not522 — APPROVED** [review]
Thank you! LGTM.

File: optuna/storages/_redis.py
Comment (keisuke-umezawa) [reply optuna/storages/_redis.py:72]: Why did not merge the suggestion?

File: optuna/storages/_redis.py
Comment (keisuke-umezawa) [inline optuna/storages/_redis.py:669]: Can you use the current time of DB side, not client side?

File: optuna/storages/_redis.py
Comment (keisuke-umezawa) [inline optuna/storages/_redis.py:650]: Can you use the current time of DB side, not client side? c.f. https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/storage.py#L1197

**Review by keisuke-umezawa — CHANGES_REQUESTED** [review]
The current code uses the time of client. But, in this case, all the clients need to use the same clock. [The RDB](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/storage.py) implementation uses DB side clock. In this case, we do not use aligned clocks in the client side.

Could you try to use the DB side clock?

File: optuna/storages/_redis.py
Comment (Turakar) [reply optuna/storages/_redis.py]: Thanks, I was not aware of this problem. Good catch!

File: optuna/storages/_redis.py
Comment (keisuke-umezawa) [reply optuna/storages/_redis.py:90]: @Turakar Sorry for again. But do you have a reason you did not merge it?

Comment (Turakar) [reply optuna/storages/_redis.py:90]: Oh, I think I lost this commit during a force push after the rebasing. That's my bad, sorry.

**Review by keisuke-umezawa — APPROVED** [review]

**Comment by keisuke-umezawa** [comment]
LGTM! Thank you for your contribution!
In my local env, it also worked.

File: optuna/storages/_redis.py
Comment (keisuke-umezawa) [inline optuna/storages/_redis.py:401]: Sorry, after adding LGTM, but it should be better to add:

```
if self.is_heartbeat_supported():
```

Comment (Turakar) [reply optuna/storages/_redis.py:401]: I think you mean `is_heartbeat_enabled()`, right?

Comment (Turakar) [reply optuna/storages/_redis.py:401]: Actually, I do not completely agree with your point. Suppose the following scenario:

1. A study is started with heartbeat enabled.
2. A trial is started.
3. The study aborts unexpectedly.
4. The study is reloaded from the storage manually on the console and the running trial is set to FAIL manually.

With the if-statement, the heartbeat would not be deleted, and as such the `_get_stale_trial_ids()` would report the trial, although it is already i state FAIL. I.e. with the if-statement you can go to some inconsistent state.

Comment (keisuke-umezawa) [reply optuna/storages/_redis.py:401]: @Turakar Thank you. I understand.
In that case, could you add the following comment?
```
# To ensure that there are no failed trials with heartbeats in the DB under any circumstances
```

After that, I will merge it.

**Review by keisuke-umezawa — APPROVED** [review]

File: optuna/storages/_redis.py
Comment (Turakar) [reply optuna/storages/_redis.py:401]: Added :slightly_smiling_face:

**Comment by Turakar** [comment]
Is it ready to merge or should I do something else?

**Comment by keisuke-umezawa** [comment]
@Turakar Sorry for being late! Meged it.

**Pull request merged**
